File: tasks/builtin/refactor-03/starter.py
"""Procedural CSV-like data processor."""


def parse_lines(raw_input):
    lines = raw_input.strip().split("\n")
    rows = []
    for line in lines:
        if line.strip():
            parts = line.split(",")
            rows.append(parts)
    return rows


def validate_rows(rows, expected_columns):
    valid_rows = []
    errors = []
    for i, row in enumerate(rows):
        if len(row) != expected_columns:
            errors.append(f"Row {i}: expected {expected_columns} columns, got {len(row)}")
        else:
            valid_rows.append(row)
    return valid_rows, errors


def parse_numeric_columns(rows, column_indices):
    parsed_rows = []
    errors = []
    for i, row in enumerate(rows):
        parsed_row = []
        has_error = False
        for j, cell in enumerate(row):
            if j in column_indices:
                try:
                    parsed_row.append(float(cell.strip()))
                except ValueError:
                    errors.append(f"Row {i}, column {j}: '{cell}' is not a valid number")
                    has_error = True
                    break
            else:
                parsed_row.append(cell.strip())
        if not has_error:
            parsed_rows.append(parsed_row)
    return parsed_rows, errors


def normalize_text_columns(rows, column_indices):
    normalized_rows = []
    for row in rows:
        normalized_row = []
        for j, cell in enumerate(row):
            if j in column_indices and isinstance(cell, str):
                normalized_row.append(cell.upper())
            else:
                normalized_row.append(cell)
        normalized_rows.append(normalized_row)
    return normalized_rows


def compute_derived_column(rows, source_col_idx, multiplier):
    result_rows = []
    for row in rows:
        new_row = row + [row[source_col_idx] * multiplier]
        result_rows.append(new_row)
    return result_rows


def format_rows_as_table(rows, column_widths):
    lines = []
    for row in rows:
        formatted_cells = []
        for i, cell in enumerate(row):
            width = column_widths[i]
            if isinstance(cell, float):
                formatted_cells.append(f"{cell:>{width}.2f}")
            else:
                formatted_cells.append(f"{str(cell):<{width}}")
        lines.append(" | ".join(formatted_cells))
    return "\n".join(lines)


def process_data(raw_input: str) -> str:
    """Process CSV-like data: parse, validate, transform, format.

    Args:
        raw_input: CSV-like input with 3 columns: name,category,value

    Returns:
        Formatted table string with 4 columns (adds computed column)
    """
    rows = parse_lines(raw_input)
    if not rows:
        return "ERROR: No data"

    valid_rows, validation_errors = validate_rows(rows, 3)
    if validation_errors:
        return "ERROR: " + "; ".join(validation_errors)

    parsed_rows, parse_errors = parse_numeric_columns(valid_rows, {2})
    if parse_errors:
        return "ERROR: " + "; ".join(parse_errors)

    normalized_rows = normalize_text_columns(parsed_rows, {0, 1})
    computed_rows = compute_derived_column(normalized_rows, 2, 1.5)
    result = format_rows_as_table(computed_rows, [15, 12, 10, 10])

    return result

File: tasks/builtin/refactor-03-cs/Starter.cs
using System;
using System.Collections.Generic;
using System.Linq;

public static class DataProcessor
{
    public static string ProcessData(string rawInput)
    {
        var rows = ParseLines(rawInput);
        if (rows.Count == 0)
            return "ERROR: No data";

        var (validRows, validationErrors) = ValidateRows(rows, 3);
        if (validationErrors.Any())
            return "ERROR: " + string.Join("; ", validationErrors);

        var (parsedRows, parseErrors) = ParseNumericColumns(validRows, new HashSet<int> { 2 });
        if (parseErrors.Any())
            return "ERROR: " + string.Join("; ", parseErrors);

        var normalizedRows = NormalizeTextColumns(parsedRows, new HashSet<int> { 0, 1 });
        var computedRows = ComputeDerivedColumn(normalizedRows, 2, 1.5);
        var result = FormatRowsAsTable(computedRows, new[] { 15, 12, 10, 10 });

        return result;
    }

    private static List<List<string>> ParseLines(string rawInput)
    {
        var lines = rawInput.Trim().Split('\n');
        var rows = new List<List<string>>();
        foreach (var line in lines)
        {
            if (!string.IsNullOrWhiteSpace(line))
            {
                var parts = line.Split(',').ToList();
                rows.Add(parts);
            }
        }
        return rows;
    }

    private static (List<List<string>> ValidRows, List<string> Errors) ValidateRows(
        List<List<string>> rows, int expectedColumns)
    {
        var validRows = new List<List<string>>();
        var errors = new List<string>();
        for (int i = 0; i < rows.Count; i++)
        {
            if (rows[i].Count != expectedColumns)
            {
                errors.Add($"Row {i}: expected {expectedColumns} columns, got {rows[i].Count}");
            }
            else
            {
                validRows.Add(rows[i]);
            }
        }
        return (validRows, errors);
    }

    private static (List<List<object>> ParsedRows, List<string> Errors) ParseNumericColumns(
        List<List<string>> rows, HashSet<int> columnIndices)
    {
        var parsedRows = new List<List<object>>();
        var errors = new List<string>();
        for (int i = 0; i < rows.Count; i++)
        {
            var parsedRow = new List<object>();
            bool hasError = false;
            for (int j = 0; j < rows[i].Count; j++)
            {
                if (columnIndices.Contains(j))
                {
                    if (double.TryParse(rows[i][j].Trim(), out double value))
                    {
                        parsedRow.Add(value);
                    }
                    else
                    {
                        errors.Add($"Row {i}, column {j}: '{rows[i][j]}' is not a valid number");
                        hasError = true;
                        break;
                    }
                }
                else
                {
                    parsedRow.Add(rows[i][j].Trim());
                }
            }
            if (!hasError)
            {
                parsedRows.Add(parsedRow);
            }
        }
        return (parsedRows, errors);
    }

    private static List<List<object>> NormalizeTextColumns(
        List<List<object>> rows, HashSet<int> columnIndices)
    {
        var normalizedRows = new List<List<object>>();
        foreach (var row in rows)
        {
            var normalizedRow = new List<object>();
            for (int j = 0; j < row.Count; j++)
            {
                if (columnIndices.Contains(j) && row[j] is string str)
                {
                    normalizedRow.Add(str.ToUpper());
                }
                else
                {
                    normalizedRow.Add(row[j]);
                }
            }
            normalizedRows.Add(normalizedRow);
        }
        return normalizedRows;
    }

    private static List<List<object>> ComputeDerivedColumn(
        List<List<object>> rows, int sourceColIdx, double multiplier)
    {
        var resultRows = new List<List<object>>();
        foreach (var row in rows)
        {
            var newRow = new List<object>(row);
            newRow.Add((double)row[sourceColIdx] * multiplier);
            resultRows.Add(newRow);
        }
        return resultRows;
    }

    private static string FormatRowsAsTable(List<List<object>> rows, int[] columnWidths)
    {
        var lines = new List<string>();
        foreach (var row in rows)
        {
            var formattedCells = new List<string>();
            for (int i = 0; i < row.Count; i++)
            {
                int width = columnWidths[i];
                if (row[i] is double d)
                {
                    formattedCells.Add(d.ToString($"F2").PadLeft(width));
                }
                else
                {
                    formattedCells.Add(row[i].ToString()!.PadRight(width));
                }
            }
            lines.Add(string.Join(" | ", formattedCells));
        }
        return string.Join("\n", lines);
    }
}
